lib/services/share.dart

```diff
@@ -0,0 +1,155 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:Okuna/plugins/share/share.dart';
+import 'package:Okuna/services/localization.dart';
+import 'package:Okuna/services/toast.dart';
+import 'package:Okuna/services/validation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+
+import 'media.dart';
+
+class ShareService {
+  static const _stream = const EventChannel('openbook.social/receive_share');
+
+  ToastService _toastService;
+  MediaService _mediaService;
+  ValidationService _validationService;
+  LocalizationService _localizationService;
+
+  StreamSubscription _shareReceiveSubscription;
+  List<Share> _shareQueue;
+  List<Future<bool> Function({String text, File image, File video})> _subscribers;
+
+  BuildContext _context;
+
+  ShareService() {
+    _shareQueue = [];
+    _subscribers = [];
+
+    if (Platform.isAndroid) {
+      if (_shareReceiveSubscription == null) {
+        _shareReceiveSubscription =
+            _stream.receiveBroadcastStream().listen(_onReceiveShare);
+      }
+    }
+  }
+
+  void setToastService(ToastService toastService) {
+    _toastService = toastService;
+  }
+
+  void setValidationService(ValidationService validationService) {
+    _validationService = validationService;
+  }
+
+  void setLocalizationService(LocalizationService localizationService) {
+    _localizationService = localizationService;
+  }
+
+  void setMediaService(MediaService mediaService) {
+    _mediaService = mediaService;
+  }
+
+  void setContext(BuildContext context) {
+    _context = context;
+  }
+
+  /// Subscribe to share events.
+  ///
+  /// [onShare] should return [true] if it consumes the share. If [false] is
```

Comment:
Cool! So is this how we get the share first in the modal and not double handle it on the home page? If so, very clever!

Reply:
The last subscriber in the list is sent the share. If it doesn't "consume" it, we pass it to the next subscriber, etc.

The "return true to consume" pattern is probably overkill at this stage since everything that subscribes to a share consumes it, but I'm so used to event systems working like this that I default to it. 😅

Reply:
It's cool though. Would love for all event systems to be designed this way.

Reply:
That's one of my biggest issues with Flutter, how extremely complicated the gesture system is if you want a simple "can I handle this gesture right now? no? then pass it to the next guy in the list".

Reply:
Good stuff @Komposten !